lib/xhttp.ex

```diff
@@ -69,10 +231,74 @@ defmodule XHTTP do
+  Checks whether the connection is open.
+
+  This function returns `true` if the connection is open, `false` otherwise. It should
+  be used to check that a connection is open before sending requests or performing
+  operations that involve talking to the server.
+
+  If a connection is not open, it has become useless and you should get rid of it.
+  If you still need a connection to the server, start a new connection with `connect/4`.
+
+  ## Examples
+
+      {:ok, conn} = XHTTP.connect(:http, "httpbin.org", 80)
+      XHTTP.open?(conn)
+      #=> true
+
+  """
   @impl true
   @spec open?(t()) :: boolean()
   def open?(conn), do: conn_module(conn).open?(conn)
 
+  @doc """
+  Sends a request to the connected server.
+
+  This function sends a new request to the server that `conn` is connected to.
+  `method` is a string representing the method for the request, such as `"GET"`
+  or `"POST"`. `path` is the path on the host to send the request to. `headers`
+  is a list of request headers in the form `{header_name, header_value}` with
+  `header_name` and `header_value` being strings. `body` can have one of three
+  values:
+
+    * `nil` - no body is sent with the request. This is the default value.
+
+    * iodata - the body to send for the request.
+
+    * `:stream` - when the value of the body is `:stream` the request
+      body can be streamed on the connection. See `stream_request_body/3`.
+      In HTTP/1.1, you can't open a request if the body of another request is
+      streaming.
+
+  If the request is sent correctly, this function returns `{:ok, conn, request_ref}`.
+  `conn` is an updated connection that should be stored over the old connection.
+  `request_ref` is a unique reference that can be used to match on responses for this
+  request that are returned by `stream/2`. See `stream/2` for more information.
+
+  If there's an error with sending the request, `{:error, conn, reason}` is returned.
+  `reason` is the cause of the error. `conn` is an updated connection. It's important
+  to store the returned connection over the old connection in case of errors too, because
+  the state of the connection might change when there are errors as well. An error when
+  sending a request **does not** necessarily mean that the connection is closed. Use
+  `open?/1` to verify that the connection is open.
+
+  Requests can be pipelined so the full response does not have to received
+  before the next request can be sent. It is up to users to verify that the
+  server supports pipelining and that the request is safe to pipeline.
+
+  In HTTP/1.1, you can't open a request if the body of another request is streaming.
+  See `XHTTP1.request/5` for more information.
+
+  For a quick discussion on HTTP/2 streams and requests, see the `XHTTP2` module and
+  `XHTTP2.request/5`.
+
+  ## Examples
+
+      XHTTP.request(conn, "GET", "/", _headers = [])
+      XHTTP.request(conn, "POST", "/path", [{"Content-Type", "application/json"}], "{}")
```

Comment:
```suggestion
      XHTTP.request(conn, "POST", "/path", [{"content-type", "application/json"}], "{}")
```

I was thinking we should use lower case headers to promote that, not super important...

Reply:
Definitely agree on that.